{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gbouras13/plassembler/blob/main/run_plassembler.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QGd2GEI3N-02"
      },
      "source": [
        "##Plassembler\n",
        "\n",
        "[plassembler](https://github.com/gbouras13/plassembler) is a program that is designed for automated & fast assembly of plasmids in  bacterial genomes that have been hybrid sequenced with long read & paired-end short read sequencing. It was originally designed for Oxford Nanopore Technologies long reads, but it will also work with Pacbio reads. As of v1.3.0, it also works well for long-read only assembled genomes.\n",
        "\n",
        "The full documentation for Plassembler can be found [here](https://plassembler.readthedocs.io/en/latest).\n",
        "\n",
        "**To run the code cells, press the play buttons on the top left of each block**\n",
        "\n",
        "Main Instructions\n",
        "\n",
        "* Cells 1 and 2 installs plassembler and download the databases. These must be run first.\n",
        "* Once they have been run, you can run either Cell 3, Cell 4 or both as many times as you wish.\n",
        "* To run `plassembler run` (if you have both long- and short-reads), run Cell 3.\n",
        "* To run `plassembler long` (if you have only long-reads), run Cell 4.\n",
        "\n",
        "Other instructions\n",
        "\n",
        "* Please make sure you change the runtime to CPU (GPU is not required).\n",
        "* To do this, go to the top toolbar, then to Runtime -> Change runtime type -> Hardware accelerator\n",
        "* You may want to upload your FASTQ files first as this takes a while for large files\n",
        "* Click on the folder icon to the left and use file upload button (with the upwards facing arrow)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ii39RG8eOZUx",
        "outputId": "aa815c67-14b1-4ff2-a6e9-bb64ca5a473e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3.11\n",
            "CPU times: user 2.61 ms, sys: 0 ns, total: 2.61 ms\n",
            "Wall time: 4.48 ms\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "32512"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "#@title 1. Install plassembler\n",
        "\n",
        "#@markdown This cell installs plassembler.\n",
        "\n",
        "%%time\n",
        "import os\n",
        "from sys import version_info\n",
        "python_version = f\"{version_info.major}.{version_info.minor}\"\n",
        "PYTHON_VERSION = python_version\n",
        "PLASSEMBLER_VERSION = \"1.6.2\"\n",
        "\n",
        "print(PYTHON_VERSION)\n",
        "\n",
        "if not os.path.isfile(\"CONDA_READY\"):\n",
        "  print(\"installing conda...\")\n",
        "  os.system(\"wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\")\n",
        "  os.system(\"bash Miniconda3-latest-Linux-x86_64.sh -bfp /usr/local\")\n",
        "  os.system(\"mamba config --set auto_update_conda false\")\n",
        "  os.system(\"touch CONDA_READY\")\n",
        "\n",
        "if not os.path.isfile(\"PLASSEMBLER_READY\"):\n",
        "  print(\"installing plassembler ...\")\n",
        "  os.system(f\"conda create -n plassemblerENV -y -c conda-forge -c bioconda python=3.9 plassembler==1.6.2 unicycler==0.5.0\")\n",
        "  os.system(\"touch PLASSEMBLER_READY\")\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tfltpbZ_QLfZ",
        "outputId": "2806708d-d30f-45d1-a778-9a25a3f9e352"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading plassembler database. This will take some time. Please be patient :)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0"
            ]
          },
          "metadata": {},
          "execution_count": 2
        }
      ],
      "source": [
        "#@title 2. Download plassembler database\n",
        "\n",
        "#@markdown This cell downloads the plassembler database.\n",
        "#@markdown It will take some time (5-10 mins). Please be patient.\n",
        "%%bash\n",
        "source activate /usr/local/envs/plassemblerENV\n",
        "python\n",
        "print(\"Downloading plassembler database. This will take some time. Please be patient :)\")\n",
        "os.system(\"/usr/local/envs/plassemblerENV/bin/plassembler download -d plassembler_db\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Sjdpu6R-Kig9",
        "outputId": "630f7e1b-28fe-4846-c5a9-16a238872fd3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input file input_fastq.gz exists\n",
            "Input file input_R1.fastq.gz exists\n",
            "Input file input_R2.fastq.gz exists\n",
            "Running plassembler run\n",
            "plassembler run completed successfully.\n",
            "Your output is in plassembler_output.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to plassembler_output.zip\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "2025-01-16 06:09:00.342 | INFO     | plassembler:begin_plassembler:100 - You are using Plassembler version 1.6.2\n",
            "2025-01-16 06:09:00.342 | INFO     | plassembler:begin_plassembler:101 - Repository homepage is https://github.com/gbouras13/plassembler\n",
            "2025-01-16 06:09:00.342 | INFO     | plassembler:begin_plassembler:102 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
            "2025-01-16 06:09:00.343 | INFO     | plassembler:run:405 - Database directory is plassembler_db\n",
            "2025-01-16 06:09:00.343 | INFO     | plassembler:run:406 - Longreads file is input_fastq.gz\n",
            "2025-01-16 06:09:00.343 | INFO     | plassembler:run:407 - R1 fasta file is input_R1.fastq.gz\n",
            "2025-01-16 06:09:00.343 | INFO     | plassembler:run:408 - R2 fasta file is input_R2.fastq.gz\n",
            "2025-01-16 06:09:00.343 | INFO     | plassembler:run:409 - Chromosome length threshold is 50000\n",
            "2025-01-16 06:09:00.344 | INFO     | plassembler:run:410 - Output directory is plassembler_output\n",
            "2025-01-16 06:09:00.344 | INFO     | plassembler:run:411 - Min long read length is 1000\n",
            "2025-01-16 06:09:00.344 | INFO     | plassembler:run:412 - Min long read quality is 9\n",
            "2025-01-16 06:09:00.344 | INFO     | plassembler:run:413 - Thread count is 2\n",
            "2025-01-16 06:09:00.344 | INFO     | plassembler:run:414 - --force is True\n",
            "2025-01-16 06:09:00.345 | INFO     | plassembler:run:415 - --skip_qc is False\n",
            "2025-01-16 06:09:00.345 | INFO     | plassembler:run:416 - --raw_flag is False\n",
            "2025-01-16 06:09:00.345 | INFO     | plassembler:run:417 - --pacbio_model is nothing\n",
            "2025-01-16 06:09:00.345 | INFO     | plassembler:run:418 - --keep_fastqs is False\n",
            "2025-01-16 06:09:00.345 | INFO     | plassembler:run:419 - --keep_chromosome is False\n",
            "2025-01-16 06:09:00.345 | INFO     | plassembler:run:420 - --flye_directory is nothing\n",
            "2025-01-16 06:09:00.346 | INFO     | plassembler:run:421 - --flye_assembly is nothing\n",
            "2025-01-16 06:09:00.346 | INFO     | plassembler:run:422 - --flye_info is nothing\n",
            "2025-01-16 06:09:00.346 | INFO     | plassembler:run:423 - --no_chromosome is False\n",
            "2025-01-16 06:09:00.346 | INFO     | plassembler:run:424 - --use_raven is False\n",
            "2025-01-16 06:09:00.346 | INFO     | plassembler:run:425 - --depth_filter is 0.25\n",
            "2025-01-16 06:09:00.347 | INFO     | plassembler:run:426 - --unicycler_options is None\n",
            "2025-01-16 06:09:00.347 | INFO     | plassembler:run:427 - --spades_options is None\n",
            "2025-01-16 06:09:00.347 | INFO     | plassembler:run:431 - Checking dependencies\n",
            "2025-01-16 06:09:00.456 | INFO     | plassembler.utils.input_commands:check_dependencies:199 - Flye version found is v2.9.5-b1801.\n",
            "2025-01-16 06:09:00.457 | INFO     | plassembler.utils.input_commands:check_dependencies:209 - Flye version is ok.\n",
            "2025-01-16 06:09:00.464 | INFO     | plassembler.utils.input_commands:check_dependencies:218 - Raven v1.8.3 found.\n",
            "2025-01-16 06:09:00.465 | INFO     | plassembler.utils.input_commands:check_dependencies:220 - Raven version is ok.\n",
            "2025-01-16 06:09:00.564 | INFO     | plassembler.utils.input_commands:check_dependencies:242 - Unicycler version found is v0.5.0.\n",
            "2025-01-16 06:09:00.565 | INFO     | plassembler.utils.input_commands:check_dependencies:255 - Unicycler version is ok.\n",
            "2025-01-16 06:09:00.928 | INFO     | plassembler.utils.input_commands:check_dependencies:265 - SPAdes v3.15.5 found.\n",
            "2025-01-16 06:09:00.943 | INFO     | plassembler.utils.input_commands:check_dependencies:278 - Samtools v1.21 found.\n",
            "2025-01-16 06:09:00.950 | INFO     | plassembler.utils.input_commands:check_dependencies:289 - minimap2 v2.28-r1209 found.\n",
            "2025-01-16 06:09:00.959 | INFO     | plassembler.utils.input_commands:check_dependencies:300 - fastp v0.24.0 found.\n",
            "2025-01-16 06:09:00.967 | INFO     | plassembler.utils.input_commands:check_dependencies:311 - chopper v0.8.0 found.\n",
            "2025-01-16 06:09:00.980 | INFO     | plassembler.utils.input_commands:check_dependencies:326 - mash v2.3 found.\n",
            "2025-01-16 06:09:01.182 | INFO     | plassembler.utils.input_commands:check_dependencies:337 - canu v2.2 found.\n",
            "2025-01-16 06:09:02.021 | INFO     | plassembler.utils.input_commands:check_dependencies:348 - dnaapler v1.1.0 found.\n",
            "2025-01-16 06:09:02.657 | INFO     | plassembler.utils.input_commands:check_dependencies:365 - BLAST version found is v2.16.0.\n",
            "2025-01-16 06:09:02.658 | INFO     | plassembler.utils.input_commands:check_dependencies:371 - All dependencies found.\n",
            "2025-01-16 06:09:02.658 | INFO     | plassembler:run:436 - Checking database installation.\n",
            "2025-01-16 06:09:02.659 | INFO     | plassembler.utils.db:check_db_installation:52 - PLSDB Database mash sketch at plassembler_db/plsdb_2023_11_03_v2.msh exists.\n",
            "2025-01-16 06:09:02.659 | INFO     | plassembler.utils.db:check_db_installation:53 - PLSDB Database tsv metadata file at plassembler_db/plsdb_2023_11_03_v2.tsv exists.\n",
            "2025-01-16 06:09:02.659 | INFO     | plassembler.utils.db:check_db_installation:54 - PLSDB Database at plassembler_db has already been downloaded\n",
            "2025-01-16 06:09:02.659 | INFO     | plassembler:run:439 - Database successfully checked.\n",
            "2025-01-16 06:09:02.660 | INFO     | plassembler:run:442 - Checking input fastqs.\n",
            "2025-01-16 06:09:02.672 | INFO     | plassembler.utils.input_commands:validate_fastq:24 - FASTQ input_fastq.gz checked\n",
            "2025-01-16 06:09:02.673 | INFO     | plassembler.utils.input_commands:validate_fastq:24 - FASTQ input_R1.fastq.gz checked\n",
            "2025-01-16 06:09:02.673 | INFO     | plassembler.utils.input_commands:validate_fastq:24 - FASTQ input_R2.fastq.gz checked\n",
            "2025-01-16 06:09:02.674 | INFO     | plassembler:run:450 - FASTQ file input_fastq.gz compression is True\n",
            "2025-01-16 06:09:02.674 | INFO     | plassembler:run:451 - FASTQ file input_R1.fastq.gz compression is True\n",
            "2025-01-16 06:09:02.674 | INFO     | plassembler:run:452 - FASTQ file input_R2.fastq.gz compression is True\n",
            "2025-01-16 06:09:02.674 | INFO     | plassembler:run:460 - Filtering long reads with chopper\n",
            "2025-01-16 06:09:02.674 | INFO     | plassembler.utils.qc:chopper:25 - Started running chopper\n",
            "2025-01-16 06:09:04.821 | INFO     | plassembler.utils.qc:chopper:82 - Finished running chopper\n",
            "2025-01-16 06:09:04.822 | INFO     | plassembler:run:484 - Running Flye.\n",
            "2025-01-16 06:09:04.823 | INFO     | plassembler.utils.external_tools:run:58 - Started running flye --nano-hq plassembler_output/chopper_long_reads.fastq.gz --out-dir plassembler_output --threads 2 ...\n",
            "2025-01-16 06:13:38.357 | INFO     | plassembler.utils.external_tools:run:60 - Done running flye --nano-hq plassembler_output/chopper_long_reads.fastq.gz --out-dir plassembler_output --threads 2\n",
            "2025-01-16 06:13:38.358 | INFO     | plassembler:run:535 - Counting Contigs.\n",
            "2025-01-16 06:13:38.360 | INFO     | plassembler.utils.plass_class:get_contig_count:87 - Assembled 2 contigs.\n",
            "2025-01-16 06:13:38.360 | INFO     | plassembler:run:736 - More than one contig was assembled with Flye.\n",
            "2025-01-16 06:13:38.361 | INFO     | plassembler:run:737 - Extracting Chromosome.\n",
            "2025-01-16 06:13:38.406 | INFO     | plassembler:run:772 - Chromosome Identified. Plassembler will now use long and short reads to assemble plasmids accurately.\n",
            "2025-01-16 06:13:38.406 | INFO     | plassembler:run:774 - Mapping long reads.\n",
            "2025-01-16 06:13:38.407 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running minimap2 -ax map-ont -t 2 plassembler_output/flye_renamed.fasta plassembler_output/chopper_long_reads.fastq.gz ...\n",
            "2025-01-16 06:13:40.618 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running minimap2 -ax map-ont -t 2 plassembler_output/flye_renamed.fasta plassembler_output/chopper_long_reads.fastq.gz\n",
            "2025-01-16 06:13:40.618 | INFO     | plassembler:run:789 - Trimming short reads.\n",
            "2025-01-16 06:13:40.619 | INFO     | plassembler.utils.external_tools:run:58 - Started running fastp --out1 plassembler_output/trimmed_R1.fastq --out2 plassembler_output/trimmed_R2.fastq --in1 input_R1.fastq.gz --in2 input_R2.fastq.gz ...\n",
            "2025-01-16 06:13:43.785 | INFO     | plassembler.utils.external_tools:run:60 - Done running fastp --out1 plassembler_output/trimmed_R1.fastq --out2 plassembler_output/trimmed_R2.fastq --in1 input_R1.fastq.gz --in2 input_R2.fastq.gz\n",
            "2025-01-16 06:13:43.786 | INFO     | plassembler:run:793 - Mapping short reads.\n",
            "2025-01-16 06:13:43.788 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running minimap2 -ax sr -t 2 plassembler_output/flye_renamed.fasta plassembler_output/trimmed_R1.fastq plassembler_output/trimmed_R2.fastq ...\n",
            "2025-01-16 06:13:45.084 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running minimap2 -ax sr -t 2 plassembler_output/flye_renamed.fasta plassembler_output/trimmed_R1.fastq plassembler_output/trimmed_R2.fastq\n",
            "2025-01-16 06:13:45.084 | INFO     | plassembler:run:800 - Processing Sam/Bam Files and extracting Fastqs.\n",
            "2025-01-16 06:13:45.184 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running samtools view -h -@ 2 -b plassembler_output/short_read.sam ...\n",
            "2025-01-16 06:13:45.761 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running samtools view -h -@ 2 -b plassembler_output/short_read.sam\n",
            "2025-01-16 06:13:45.762 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running samtools view -b -h -@ 2 -L plassembler_output/non_chromosome.bed plassembler_output/short_read.bam ...\n",
            "2025-01-16 06:13:45.920 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running samtools view -b -h -@ 2 -L plassembler_output/non_chromosome.bed plassembler_output/short_read.bam\n",
            "2025-01-16 06:13:45.921 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running samtools view -b -h -f 4 -@ 2 plassembler_output/short_read.bam ...\n",
            "2025-01-16 06:13:45.981 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running samtools view -b -h -f 4 -@ 2 plassembler_output/short_read.bam\n",
            "2025-01-16 06:13:45.983 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running samtools view -b -h -@ 2 -L plassembler_output/chromosome.bed plassembler_output/short_read.bam ...\n",
            "2025-01-16 06:13:46.169 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running samtools view -b -h -@ 2 -L plassembler_output/chromosome.bed plassembler_output/short_read.bam\n",
            "2025-01-16 06:13:46.171 | INFO     | plassembler.utils.external_tools:run:58 - Started running samtools fastq -@ 2 plassembler_output/unmapped_bam_file.bam -1 plassembler_output/unmapped_R1.fastq -2 plassembler_output/unmapped_R2.fastq -0 /dev/null -s /dev/null -n ...\n",
            "2025-01-16 06:13:46.199 | INFO     | plassembler.utils.external_tools:run:60 - Done running samtools fastq -@ 2 plassembler_output/unmapped_bam_file.bam -1 plassembler_output/unmapped_R1.fastq -2 plassembler_output/unmapped_R2.fastq -0 /dev/null -s /dev/null -n\n",
            "2025-01-16 06:13:46.201 | INFO     | plassembler.utils.external_tools:run:58 - Started running samtools fastq -@ 2 plassembler_output/non_chromosome.bam -1 plassembler_output/mapped_non_chromosome_R1.fastq -2 plassembler_output/mapped_non_chromosome_R2.fastq -0 /dev/null -s /dev/null -n ...\n",
            "2025-01-16 06:13:46.261 | INFO     | plassembler.utils.external_tools:run:60 - Done running samtools fastq -@ 2 plassembler_output/non_chromosome.bam -1 plassembler_output/mapped_non_chromosome_R1.fastq -2 plassembler_output/mapped_non_chromosome_R2.fastq -0 /dev/null -s /dev/null -n\n",
            "2025-01-16 06:13:47.040 | INFO     | plassembler:run:817 - Running Unicycler.\n",
            "2025-01-16 06:13:47.041 | INFO     | plassembler.utils.external_tools:run:58 - Started running unicycler -1 plassembler_output/short_read_concat_R1.fastq -2 plassembler_output/short_read_concat_R2.fastq -l plassembler_output/plasmid_long.fastq -t 2 -o plassembler_output/unicycler_output ...\n",
            "2025-01-16 06:18:03.193 | INFO     | plassembler.utils.external_tools:run:60 - Done running unicycler -1 plassembler_output/short_read_concat_R1.fastq -2 plassembler_output/short_read_concat_R2.fastq -l plassembler_output/plasmid_long.fastq -t 2 -o plassembler_output/unicycler_output\n",
            "2025-01-16 06:18:03.195 | INFO     | plassembler:run:842 - Unicycler identified plasmids. Calculating Plasmid Copy Numbers.\n",
            "2025-01-16 06:18:03.198 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running minimap2 -ax map-ont -t 2 plassembler_output/combined.fasta plassembler_output/chopper_long_reads.fastq.gz ...\n",
            "2025-01-16 06:18:05.485 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running minimap2 -ax map-ont -t 2 plassembler_output/combined.fasta plassembler_output/chopper_long_reads.fastq.gz\n",
            "2025-01-16 06:18:05.486 | INFO     | plassembler.utils.external_tools:run:58 - Started running samtools sort -@ 2 plassembler_output/combined_long.sam -o plassembler_output/combined_sorted_long.bam ...\n",
            "2025-01-16 06:18:05.768 | INFO     | plassembler.utils.external_tools:run:60 - Done running samtools sort -@ 2 plassembler_output/combined_long.sam -o plassembler_output/combined_sorted_long.bam\n",
            "2025-01-16 06:18:05.770 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running minimap2 -ax sr -t 2 plassembler_output/combined.fasta plassembler_output/trimmed_R1.fastq plassembler_output/trimmed_R2.fastq ...\n",
            "2025-01-16 06:18:06.522 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running minimap2 -ax sr -t 2 plassembler_output/combined.fasta plassembler_output/trimmed_R1.fastq plassembler_output/trimmed_R2.fastq\n",
            "2025-01-16 06:18:06.523 | INFO     | plassembler.utils.external_tools:run:58 - Started running samtools sort -@ 2 plassembler_output/combined_short.sam -o plassembler_output/combined_sorted_short.bam ...\n",
            "2025-01-16 06:18:06.826 | INFO     | plassembler.utils.external_tools:run:60 - Done running samtools sort -@ 2 plassembler_output/combined_short.sam -o plassembler_output/combined_sorted_short.bam\n",
            "2025-01-16 06:18:07.884 | INFO     | plassembler:run:850 - Calculating mash distances to PLSDB.\n",
            "2025-01-16 06:18:07.885 | INFO     | plassembler.utils.external_tools:run:58 - Started running mash sketch plassembler_output/plasmids.fasta -i ...\n",
            "2025-01-16 06:18:07.946 | INFO     | plassembler.utils.external_tools:run:60 - Done running mash sketch plassembler_output/plasmids.fasta -i\n",
            "2025-01-16 06:18:07.948 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running mash dist plassembler_output/plasmids.fasta.msh plassembler_db/plsdb_2023_11_03_v2.msh -v 0.1 -d 0.1 -i ...\n",
            "2025-01-16 06:18:12.223 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running mash dist plassembler_output/plasmids.fasta.msh plassembler_db/plsdb_2023_11_03_v2.msh -v 0.1 -d 0.1 -i\n",
            "2025-01-16 06:18:13.812 | INFO     | plassembler.utils.plass_class:combine_depth_mash_tsvs:751 - Filtering contigs below depth filter: 0.25.\n",
            "2025-01-16 06:18:13.813 | INFO     | plassembler.utils.plass_class:combine_depth_mash_tsvs:753 - All plasmids whose short and long read copy numbers are both below 0.25 will be removed.\n",
            "2025-01-16 06:18:13.813 | INFO     | plassembler.utils.plass_class:combine_depth_mash_tsvs:769 - No plasmids were filtered due to low depth.\n",
            "2025-01-16 06:18:13.890 | INFO     | plassembler:end_plassembler:117 - Plassembler has finished\n",
            "2025-01-16 06:18:13.891 | INFO     | plassembler:end_plassembler:118 - Elapsed time: 553.56 seconds\n"
          ]
        }
      ],
      "source": [
        "#@title 3. Plassembler Run (Hybrid reads)\n",
        "\n",
        "#@markdown This will probably take a while (depends on your read sets - an hour or two probably: best to put it on over lunch) as the colab environment has limited resources.\n",
        "\n",
        "#@markdown First, upload your long-reads as a single input .fastq or .fastq.gz file\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use file upload button.\n",
        "\n",
        "#@markdown Once it is uploaded, write the file name in the LONG_FASTQ field on the right.\n",
        "\n",
        "#@markdown Then, upload your short-reads as 2 input .fastq or .fastq.gz files\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use file upload button.\n",
        "\n",
        "#@markdown Once they are uploaded, write the forward (R1) file name in the R1_FASTQ field on the right and the the reverse (R2) file name in the R2_FASTQ field on the right.\n",
        "\n",
        "#@markdown Then provide a directory for plassembler's output using PLASSEMBLER_OUT_DIR.\n",
        "#@markdown The default is 'plassembler_output'.\n",
        "\n",
        "#@markdown Then provide an estimated chromosome size (as an integer) name using CHROMOSOME.\n",
        "#@markdown The default is 1000000.\n",
        "\n",
        "#@markdown You can also provide a min_length for QC filtering the long read data with MIN_LENGTH.\n",
        "#@markdown If you provide nothing it will default to 1000.\n",
        "\n",
        "#@markdown You can also provide a min_quality for QC filtering the long read data with MIN_QUALITY.\n",
        "#@markdown If you provide nothing it will default to 9.\n",
        "\n",
        "#@markdown You can click SKIP_QC to turn off qc (fastp and filtlong).\n",
        "#@markdown By default it is False.\n",
        "\n",
        "#@markdown You can click RAW to  pass --nano-raw for Flye.  Designed for Guppy fast\n",
        "#@markdown configuration reads.  By default, Flye will assume\n",
        "#@markdown SUP or HAC reads and use --nano-hq.\n",
        "\n",
        "#@markdown If you have pacbio reads, please change PACBIO_MODEL\n",
        "#@markdown from 'none' to one of 'pacbio-hifi', 'pacbio-corr', 'pacbio-raw'. Use pacbio-raw for\n",
        "#@markdown  PacBio regular CLR reads (<20 percent error),\n",
        "#@markdown pacbio-corr for PacBio reads that were corrected with other methods (<3 percent error) or pacbio-\n",
        "#@markdown  hifi for PacBio HiFi reads (<1 percent error).\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory.\n",
        "#@markdown This may be useful if your earlier pharokka run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of `plassembler run` will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PLASSEMBLER_OUT_DIR.zip, where PLASSEMBLER_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "%%bash\n",
        "source activate /usr/local/envs/plassemblerENV\n",
        "python\n",
        "import os\n",
        "import sys\n",
        "import subprocess\n",
        "import zipfile\n",
        "\n",
        "LONG_FASTQ = '' #@param {type:\"string\"}\n",
        "R1_FASTQ = '' #@param {type:\"string\"}\n",
        "R2_FASTQ = '' #@param {type:\"string\"}\n",
        "THREADS = \"2\"\n",
        "\n",
        "if os.path.exists(LONG_FASTQ):\n",
        "    print(f\"Input file {LONG_FASTQ} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {LONG_FASTQ} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "if os.path.exists(R1_FASTQ):\n",
        "    print(f\"Input file {R1_FASTQ} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {R1_FASTQ} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "if os.path.exists(R2_FASTQ):\n",
        "    print(f\"Input file {R2_FASTQ} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {R2_FASTQ} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "PLASSEMBLER_OUT_DIR = 'plassembler_output'  #@param {type:\"string\"}\n",
        "CHROMOSOME = 1000000  #@param {type:\"integer\"}\n",
        "MIN_LENGTH = 1000  #@param {type:\"integer\"}\n",
        "MIN_QUALITY = 9  #@param {type:\"integer\"}\n",
        "SKIP_QC = False  #@param {type:\"boolean\"}\n",
        "RAW = False  #@param {type:\"boolean\"}\n",
        "PACBIO_MODEL = 'none'  #@param {type:\"string\"}\n",
        "allowed_gene_predictors = ['none', 'pacbio-hifi', 'pacbio-corr', 'pacbio-raw']\n",
        "# Check if the input parameter is valid\n",
        "if PACBIO_MODEL.lower() not in allowed_gene_predictors:\n",
        "    raise ValueError(\"Invalid PACBIO_MODEL. Please choose from: 'none', 'pacbio-hifi', 'pacbio-corr', 'pacbio-raw'.\")\n",
        "\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "# Construct the command\n",
        "command = f\"/usr/local/envs/plassemblerENV/bin/plassembler run -d plassembler_db -c {CHROMOSOME} -l {LONG_FASTQ} -1 {R1_FASTQ} -2 {R2_FASTQ} -o {PLASSEMBLER_OUT_DIR} -t {THREADS} --min_length {MIN_LENGTH} --min_quality {MIN_QUALITY}\"\n",
        "\n",
        "if SKIP_QC is True:\n",
        "  command = f\"{command} --skip_qc\"\n",
        "\n",
        "if RAW is True:\n",
        "  command = f\"{command} -r\"\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "if PACBIO_MODEL != 'none':\n",
        "  command = f\"{command}  --pacbio_model {PACBIO_MODEL}\"\n",
        "\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running plassembler run\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"plassembler run completed successfully.\")\n",
        "    print(f\"Your output is in {PLASSEMBLER_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PLASSEMBLER_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PLASSEMBLER_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PLASSEMBLER_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "id": "9QfjP3q-Q04f",
        "outputId": "f7606120-77fb-4b7d-8864-3c52940ec88b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input file input_fastq.gz exists\n",
            "Running plassembler long\n",
            "plassembler long completed successfully.\n",
            "Your output is in plassembler_output.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to plassembler_output.zip\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "2025-01-16 06:22:17.399 | INFO     | plassembler:begin_plassembler:100 - You are using Plassembler version 1.6.2\n",
            "2025-01-16 06:22:17.400 | INFO     | plassembler:begin_plassembler:101 - Repository homepage is https://github.com/gbouras13/plassembler\n",
            "2025-01-16 06:22:17.400 | INFO     | plassembler:begin_plassembler:102 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
            "2025-01-16 06:22:17.400 | INFO     | plassembler:long:1294 - Database directory is plassembler_db\n",
            "2025-01-16 06:22:17.400 | INFO     | plassembler:long:1295 - Longreads file is input_fastq.gz\n",
            "2025-01-16 06:22:17.400 | INFO     | plassembler:long:1296 - Chromosome length threshold is 50000\n",
            "2025-01-16 06:22:17.401 | INFO     | plassembler:long:1297 - Output directory is plassembler_output\n",
            "2025-01-16 06:22:17.401 | INFO     | plassembler:long:1298 - Min long read length is 1000\n",
            "2025-01-16 06:22:17.401 | INFO     | plassembler:long:1299 - Min long read quality is 9\n",
            "2025-01-16 06:22:17.401 | INFO     | plassembler:long:1300 - Thread count is 2\n",
            "2025-01-16 06:22:17.401 | INFO     | plassembler:long:1301 - --force is True\n",
            "2025-01-16 06:22:17.402 | INFO     | plassembler:long:1302 - --skip_qc is False\n",
            "2025-01-16 06:22:17.402 | INFO     | plassembler:long:1303 - --raw_flag is False\n",
            "2025-01-16 06:22:17.402 | INFO     | plassembler:long:1304 - --pacbio_model is nothing\n",
            "2025-01-16 06:22:17.402 | INFO     | plassembler:long:1305 - --keep_chromosome is False\n",
            "2025-01-16 06:22:17.402 | INFO     | plassembler:long:1306 - --flye_directory is nothing\n",
            "2025-01-16 06:22:17.402 | INFO     | plassembler:long:1307 - --flye_assembly is nothing\n",
            "2025-01-16 06:22:17.403 | INFO     | plassembler:long:1308 - --flye_info is nothing\n",
            "2025-01-16 06:22:17.403 | INFO     | plassembler:long:1309 - --corrected_error_rate is 0.12\n",
            "2025-01-16 06:22:17.403 | INFO     | plassembler:long:1310 - --no_chromosome is False\n",
            "2025-01-16 06:22:17.403 | INFO     | plassembler:long:1311 - --depth_filter is 0.25\n",
            "2025-01-16 06:22:17.403 | INFO     | plassembler:long:1312 - --unicycler_options is None\n",
            "2025-01-16 06:22:17.403 | INFO     | plassembler:long:1313 - --spades_options is None\n",
            "2025-01-16 06:22:17.404 | INFO     | plassembler:long:1317 - Checking dependencies\n",
            "2025-01-16 06:22:17.491 | INFO     | plassembler.utils.input_commands:check_dependencies:199 - Flye version found is v2.9.5-b1801.\n",
            "2025-01-16 06:22:17.492 | INFO     | plassembler.utils.input_commands:check_dependencies:209 - Flye version is ok.\n",
            "2025-01-16 06:22:17.501 | INFO     | plassembler.utils.input_commands:check_dependencies:218 - Raven v1.8.3 found.\n",
            "2025-01-16 06:22:17.501 | INFO     | plassembler.utils.input_commands:check_dependencies:220 - Raven version is ok.\n",
            "2025-01-16 06:22:17.614 | INFO     | plassembler.utils.input_commands:check_dependencies:242 - Unicycler version found is v0.5.0.\n",
            "2025-01-16 06:22:17.614 | INFO     | plassembler.utils.input_commands:check_dependencies:255 - Unicycler version is ok.\n",
            "2025-01-16 06:22:17.913 | INFO     | plassembler.utils.input_commands:check_dependencies:265 - SPAdes v3.15.5 found.\n",
            "2025-01-16 06:22:17.960 | INFO     | plassembler.utils.input_commands:check_dependencies:278 - Samtools v1.21 found.\n",
            "2025-01-16 06:22:17.979 | INFO     | plassembler.utils.input_commands:check_dependencies:289 - minimap2 v2.28-r1209 found.\n",
            "2025-01-16 06:22:18.007 | INFO     | plassembler.utils.input_commands:check_dependencies:300 - fastp v0.24.0 found.\n",
            "2025-01-16 06:22:18.022 | INFO     | plassembler.utils.input_commands:check_dependencies:311 - chopper v0.8.0 found.\n",
            "2025-01-16 06:22:18.041 | INFO     | plassembler.utils.input_commands:check_dependencies:326 - mash v2.3 found.\n",
            "2025-01-16 06:22:18.199 | INFO     | plassembler.utils.input_commands:check_dependencies:337 - canu v2.2 found.\n",
            "2025-01-16 06:22:18.961 | INFO     | plassembler.utils.input_commands:check_dependencies:348 - dnaapler v1.1.0 found.\n",
            "2025-01-16 06:22:19.155 | INFO     | plassembler.utils.input_commands:check_dependencies:365 - BLAST version found is v2.16.0.\n",
            "2025-01-16 06:22:19.156 | INFO     | plassembler.utils.input_commands:check_dependencies:371 - All dependencies found.\n",
            "2025-01-16 06:22:19.156 | INFO     | plassembler:long:1321 - Checking database installation.\n",
            "2025-01-16 06:22:19.157 | INFO     | plassembler.utils.db:check_db_installation:52 - PLSDB Database mash sketch at plassembler_db/plsdb_2023_11_03_v2.msh exists.\n",
            "2025-01-16 06:22:19.157 | INFO     | plassembler.utils.db:check_db_installation:53 - PLSDB Database tsv metadata file at plassembler_db/plsdb_2023_11_03_v2.tsv exists.\n",
            "2025-01-16 06:22:19.157 | INFO     | plassembler.utils.db:check_db_installation:54 - PLSDB Database at plassembler_db has already been downloaded\n",
            "2025-01-16 06:22:19.157 | INFO     | plassembler:long:1324 - Database successfully checked.\n",
            "2025-01-16 06:22:19.157 | INFO     | plassembler:long:1327 - Checking input fastqs.\n",
            "2025-01-16 06:22:19.211 | INFO     | plassembler.utils.input_commands:validate_fastq:24 - FASTQ input_fastq.gz checked\n",
            "2025-01-16 06:22:19.211 | INFO     | plassembler:long:1350 - Filtering long reads with chopper\n",
            "2025-01-16 06:22:19.212 | INFO     | plassembler.utils.qc:chopper:25 - Started running chopper\n",
            "2025-01-16 06:22:21.482 | INFO     | plassembler.utils.qc:chopper:82 - Finished running chopper\n",
            "2025-01-16 06:22:21.483 | INFO     | plassembler:long:1373 - Running Flye.\n",
            "2025-01-16 06:22:21.484 | INFO     | plassembler.utils.external_tools:run:58 - Started running flye --nano-hq plassembler_output/chopper_long_reads.fastq.gz --out-dir plassembler_output --threads 2 ...\n",
            "2025-01-16 06:27:07.739 | INFO     | plassembler.utils.external_tools:run:60 - Done running flye --nano-hq plassembler_output/chopper_long_reads.fastq.gz --out-dir plassembler_output --threads 2\n",
            "2025-01-16 06:27:07.740 | INFO     | plassembler:long:1421 - Counting Contigs.\n",
            "2025-01-16 06:27:07.742 | INFO     | plassembler.utils.plass_class:get_contig_count:87 - Assembled 2 contigs.\n",
            "2025-01-16 06:27:07.751 | INFO     | plassembler:long:1466 - Mapping long reads.\n",
            "2025-01-16 06:27:07.752 | INFO     | plassembler.utils.external_tools:run_to_stdout:67 - Started running minimap2 -ax map-ont -t 2 plassembler_output/flye_renamed.fasta plassembler_output/chopper_long_reads.fastq.gz ...\n",
            "2025-01-16 06:27:09.941 | INFO     | plassembler.utils.external_tools:run_to_stdout:69 - Done running minimap2 -ax map-ont -t 2 plassembler_output/flye_renamed.fasta plassembler_output/chopper_long_reads.fastq.gz\n",
            "2025-01-16 06:27:09.942 | INFO     | plassembler:long:1475 - Processing Sam/Bam Files and extracting Fastqs.\n",
            "2025-01-16 06:27:10.030 | INFO     | plassembler:long:1539 - Removing junk low entropy repeat reads.\n",
            "2025-01-16 06:27:12.432 | INFO     | plassembler:long:1547 - Correcting reads with canu prior to running Unicycler.\n",
            "2025-01-16 06:27:12.433 | INFO     | plassembler.utils.external_tools:run:58 - Started running canu -correct -p canu -d plassembler_output/canu genomeSize=0.04401m maxInputCoverage=50 stopOnLowCoverage=1 maxThreads=2 -nanopore correctedErrorRate=0.12 plassembler_output/plasmid_long_entropy_filtered.fastq ...\n",
            "2025-01-16 06:27:13.433 | WARNING  | plassembler.utils.external_tools:run_tool:129 - Canu failed to correct any reads.\n",
            "2025-01-16 06:27:13.433 | WARNING  | plassembler.utils.external_tools:run_tool:130 - This probably means there is low depth, don't be too concerned.\n",
            "2025-01-16 06:27:13.433 | WARNING  | plassembler.utils.external_tools:run_tool:133 - If you are concerned, check stdout log file: plassembler_output/logs/canu__correct_6a32ab91331bb18902d5a111599b0753239ff992eb2583498db8d9b049be78ae.out and stderr log file: plassembler_output/logs/canu__correct_6a32ab91331bb18902d5a111599b0753239ff992eb2583498db8d9b049be78ae.err.\n",
            "2025-01-16 06:27:13.434 | WARNING  | plassembler:long:1567 - Advancing with uncorrected reads\n",
            "2025-01-16 06:27:13.435 | INFO     | plassembler.utils.external_tools:run:58 - Started running unicycler -s plassembler_output/plasmid_long_entropy_filtered.fastq -l plassembler_output/plasmid_long_entropy_filtered.fastq -t 2 -o plassembler_output/unicycler_output ...\n",
            "2025-01-16 06:29:09.489 | WARNING  | plassembler.utils.external_tools:run_tool:113 - Unicycler has failed. This usually means that you have no plasmids. Checking.\n",
            "2025-01-16 06:29:09.491 | INFO     | plassembler:long:1605 - unicycler assembled 0 contigs.\n",
            "2025-01-16 06:29:09.491 | INFO     | plassembler:long:1608 - Your sample probably has no plasmids.\n",
            "2025-01-16 06:29:09.491 | INFO     | plassembler:long:1609 - Check the plassembler_output/plasmid_fastqs/long_plasmid.fastq file.\n",
            "2025-01-16 06:29:09.491 | INFO     | plassembler:long:1610 - If this is small (indicating few unmapped reads and therefore unicycler failed due to low depth), then your sample likely has no plasmids.\n",
            "2025-01-16 06:29:09.491 | INFO     | plassembler:long:1613 - Cleaning up intermediate files and exiting Plassembler.\n",
            "2025-01-16 06:29:09.502 | INFO     | plassembler:end_plassembler:117 - Plassembler has finished\n",
            "2025-01-16 06:29:09.503 | INFO     | plassembler:end_plassembler:118 - Elapsed time: 412.12 seconds\n"
          ]
        }
      ],
      "source": [
        "#@title 4. Plassembler Long (long-reads only)\n",
        "\n",
        "#@markdown This will probably take a while (best to put it on overnight or over lunch) as the colab environment has limited resources.\n",
        "\n",
        "#@markdown First, upload your long-reads as a single input .fastq or .fastq.gz file\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use file upload button.\n",
        "\n",
        "#@markdown Once it is uploaded, write the file name in the LONG_FASTQ field on the right.\n",
        "\n",
        "#@markdown Then provide a directory for plassembler's output using PLASSEMBLER_OUT_DIR.\n",
        "#@markdown The default is 'plassembler_output'.\n",
        "\n",
        "#@markdown Then provide an estimated chromosome size (as an integer) name using CHROMOSOME.\n",
        "#@markdown The default is 1000000.\n",
        "\n",
        "#@markdown You can also provide a min_length for QC filtering the long read data with MIN_LENGTH.\n",
        "#@markdown If you provide nothing it will default to 1000.\n",
        "\n",
        "#@markdown You can also provide a min_quality for QC filtering the long read data with MIN_QUALITY.\n",
        "#@markdown If you provide nothing it will default to 9.\n",
        "\n",
        "#@markdown You can click SKIP_QC to turn off qc (filtlong).\n",
        "#@markdown By default it is False.\n",
        "\n",
        "#@markdown You can click RAW to  pass --nano-raw for Flye.  Designed for Guppy fast\n",
        "#@markdown configuration reads.  By default, Flye will assume\n",
        "#@markdown SUP or HAC reads and use --nano-hq.\n",
        "\n",
        "#@markdown If you have pacbio reads, please change PACBIO_MODEL\n",
        "#@markdown from 'none' to one of 'pacbio-hifi', 'pacbio-corr', 'pacbio-raw'. Use pacbio-raw for\n",
        "#@markdown  PacBio regular CLR reads (<20 percent error),\n",
        "#@markdown pacbio-corr for PacBio reads that were corrected with other methods (<3 percent error) or pacbio-\n",
        "#@markdown  hifi for PacBio HiFi reads (<1 percent error).\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory.\n",
        "#@markdown This may be useful if your earlier pharokka run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of `plassembler long` will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PLASSEMBLER_OUT_DIR.zip, where PLASSEMBLER_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "%%bash\n",
        "source activate /usr/local/envs/plassemblerENV\n",
        "python\n",
        "import os\n",
        "import sys\n",
        "import subprocess\n",
        "import zipfile\n",
        "LONG_FASTQ = '' #@param {type:\"string\"}\n",
        "THREADS = \"2\"\n",
        "\n",
        "if os.path.exists(LONG_FASTQ):\n",
        "    print(f\"Input file {LONG_FASTQ} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {LONG_FASTQ} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "\n",
        "PLASSEMBLER_OUT_DIR = 'plassembler_output'  #@param {type:\"string\"}\n",
        "CHROMOSOME = 1000000  #@param {type:\"integer\"}\n",
        "MIN_LENGTH = 1000  #@param {type:\"integer\"}\n",
        "MIN_QUALITY = 9  #@param {type:\"integer\"}\n",
        "SKIP_QC = False  #@param {type:\"boolean\"}\n",
        "RAW = False  #@param {type:\"boolean\"}\n",
        "PACBIO_MODEL = 'none'  #@param {type:\"string\"}\n",
        "allowed_gene_predictors = ['none', 'pacbio-hifi', 'pacbio-corr', 'pacbio-raw']\n",
        "# Check if the input parameter is valid\n",
        "if PACBIO_MODEL.lower() not in allowed_gene_predictors:\n",
        "    raise ValueError(\"Invalid PACBIO_MODEL. Please choose from: 'none', 'pacbio-hifi', 'pacbio-corr', 'pacbio-raw'.\")\n",
        "\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "\n",
        "# Construct the command\n",
        "command = f\"/usr/local/envs/plassemblerENV/bin/plassembler long -d plassembler_db -c {CHROMOSOME} -l {LONG_FASTQ} -o {PLASSEMBLER_OUT_DIR} -t {THREADS} --min_length {MIN_LENGTH} --min_quality {MIN_QUALITY}\"\n",
        "\n",
        "if SKIP_QC is True:\n",
        "  command = f\"{command} --skip_qc\"\n",
        "\n",
        "if RAW is True:\n",
        "  command = f\"{command} -r\"\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "if PACBIO_MODEL != 'none':\n",
        "  command = f\"{command}  --pacbio_model {PACBIO_MODEL}\"\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running plassembler long\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"plassembler long completed successfully.\")\n",
        "    print(f\"Your output is in {PLASSEMBLER_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PLASSEMBLER_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PLASSEMBLER_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PLASSEMBLER_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}